### ISSUE TITLE
Getting NullReferenceException when using refresh token

### ISSUE THREAD

**Issue by darrellbooker** [DuendeArchive/IdentityModel.PCL#43]
When refreshing my token using

var result = _tokenClient.RequestRefreshTokenAsync (refreshToken).Result;

I'm getting a NullReferenceException. This doesn't happen every time after getting an AccessToken and RefreshToken but only after the tokens are hours old.  For example, I can get a AT and RT and consistently use the RT to get more AT and RT and it works fine.  Let the code sit for a few hours and try to use the RT again and it throws the error.  StackTrack is below.  I would think if it was some sort of simple expiration issue I would have got a more graceful isError=true in the TokenResponse object.

 at IdentityModel.Client.TokenClientExtensions.Merge (IdentityModel.Client.TokenClient client, System.Collections.Generic.Dictionary`2 explicitValues, System.Object extra) [0x00016] in <filename unknown>:0 
  at IdentityModel.Client.TokenClientExtensions.RequestRefreshTokenAsync (IdentityModel.Client.TokenClient client, System.String refreshToken, System.Object extra, CancellationToken cancellationToken) [0x00024] in <filename unknown>:0 
  at SellMore.Functions.RefreshToken (System.String refreshToken) [0x0001d] in /Users/darrell/Mono Projects/SellMore/SellMoreApp/SellMore/SellMore/Classes/Functions.cs:41

**Comment by leastprivilege** [comment]
This seems to happen in the Merge method..

Maybe you are passing some null value in?

**Comment by darrellbooker** [comment]
ok so it was a stupid mistake.  my "_tokenClient" variable was null.  Now that's fixed I'm getting "invalid_grant" when refreshing the token.  At the time of the writing the token is only about a day old. Does the invalid_grant means its expired or something else?  Below are my server settings

`AccessTokenType = AccessTokenType.Reference,
RefreshTokenUsage = TokenUsage.OneTimeOnly,
RefreshTokenExpiration = TokenExpiration.Sliding,`

With these settings the token should be around for at least the default 30 days correct?

**Comment by brockallen** [comment]
Check the logs to know for certain

**Comment by darrellbooker** [comment]
Says refresh token is invalid....million dollar question is does that mean the refreshToken was expired or does it mean it was just flat out the wrong token being provided?

2016-03-30 19:01:14.482 -04:00 [Warning] Refresh token is invalid
 {
  "ClientId": "sellMoreApp",
  "ClientName": "SellMore app",
  "GrantType": "refresh_token",
  "RefreshToken": "4a4a7f87a3d2eea66416447a80c7ad4f",
  "Raw": {
    "grant_type": "refresh_token",
    "refresh_token": "4a4a7f87a3d2eea66416447a80c7ad4f"
  }
}
2016-03-30 19:01:14.484 -04:00 [Information] End token request
2016-03-30 19:01:14.489 -04:00 [Information] Returning error: invalid_grant

**Comment by leastprivilege** [comment]
It means that the refresh token handle passed in cannot be found in the store. Either it got purged or it is an invalid handle.

**Comment by leastprivilege** [comment]
everything else would be a new issue - in a different repo. closing.
